pkg/assetfetcher/http.go

```diff
@@ -55,126 +54,78 @@ type NexposeAssetFetcher struct {
 	StatFn domain.StatFn
 }
 
-// FetchAssets gets all the assets for a given site ID from Nexpose by calling `/api/3/sites/{id}/assets`.
-// This function is asynchronous, which means you can start listening to the AssetEvent and error channels immediately.
-// Assets will be added to the AssetEvent channel as they're returned from Nexpose
-// and errors will be added to the error channel if there's an error fetching
-// or reading the asset. It's the responsibility of the caller to check if a channel is closed before reading from it.
-func (c *NexposeAssetFetcher) FetchAssets(ctx context.Context, siteID string, scanID string) (<-chan domain.AssetEvent, <-chan error) {
-	errChan := make(chan error, 1)
-	defer close(errChan)
-
-	stater := c.StatFn(ctx)
-
+// FetchAssets gets all the assets for a given site ID from Nexpose by calling `/api/3/sites/{id}/assets`
+func (c *NexposeAssetFetcher) FetchAssets(ctx context.Context, siteID string) ([]domain.Asset, error) {
 	// have to page through to get all the assets, start with page 0
 	// make the first call to Nexpose to get the total number of pages we'll
 	var currentPage = 0
-	req := c.newNexposeSiteAssetsRequest(siteID, currentPage)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	res, err := c.HTTPClient.Do(req.WithContext(ctx))
-	if err != nil {
-		errChan <- &NexposeHTTPRequestError{err, req.URL.String()}
-		return nil, errChan
-	}
-	defer res.Body.Close()
-	respBody, err := ioutil.ReadAll(res.Body)
-	if err != nil {
-		errChan <- &ErrorReadingNexposeResponse{err, req.URL.String()}
-		return nil, errChan
+	firstPageOfAssets, initialErr := c.fetchNexposeSiteAssetsPage(ctx, currentPage, siteID)
+	if initialErr != nil {
+		return []domain.Asset{}, &domain.ErrorFetchingAssets{Inner: initialErr, Page: 0, SiteID: siteID}
 	}
 
-	if res.StatusCode != http.StatusOK {
-		errChan <- &ErrorFetchingAssets{Inner: fmt.Errorf("unexpected response from nexpose: %d",
-			res.StatusCode)}
-		return nil, errChan
-	}
+	totalPages := firstPageOfAssets.Page.TotalPages
+	totalAssets := firstPageOfAssets.Resources
 
-	var siteAssetResp SiteAssetsResponse
-	if err := json.Unmarshal(respBody, &siteAssetResp); err != nil {
-		errChan <- &ErrorParsingJSONResponse{err, req.URL.String()}
-		return nil, errChan
-	}
-
-	stater.Count("totalassets", float64(siteAssetResp.Page.TotalResources), fmt.Sprintf("site:%s", siteID))
-	pagedAssetChan := make(chan domain.AssetEvent, siteAssetResp.Page.TotalResources)
-	pagedErrChan := make(chan error, siteAssetResp.Page.TotalResources)
-	for _, asset := range siteAssetResp.Resources {
-		scanTime, err := asset.GetScanTime(scanID)
-		if err != nil {
-			pagedErrChan <- err
-			continue
-		}
-		assetEvent, err := asset.AssetPayloadToAssetEvent(scanTime)
-		if err != nil {
-			pagedErrChan <- err
-		} else {
-			pagedAssetChan <- assetEvent
-		}
```

Comment:
❤️ All that red is a sign you're doing the right thing. 😄

Reply:
yea i question the ratio of addition to deletion of code on the PR...